### PR TITLE
feat: allow user customizable prompts for reusable logic

### PR DESCRIPTION
**Pull request by TerminalFi**
Copilot uses GPT in the background. Essentially meaning you can create prompts in the chat panel. This just allows users to add re-usable prompts that can be triggered by the same logic as the build in prompts

### CODE REVIEW

**Comment by jfcherng** [comment]
I feel it's just weird to have `{{ sel[0] }}` in the prompt so I change it to [`17b7b1a` (#194)](https://github.com/TerminalFi/LSP-copilot/pull/194/commits/17b7b1a18e813400f5d93e25736b4f6117959551). But it also "always" appends the selected code to the end.

@TerminalFi What do you think?

**Comment by TerminalFi** [comment]
> I feel it's just weird to have `{{ sel[0] }}` in the prompt so I change it to [`17b7b1a` (#194)](https://github.com/TerminalFi/LSP-copilot/pull/194/commits/17b7b1a18e813400f5d93e25736b4f6117959551). But it also "always" appends the selected code to the end.
> 
> 
> 
> At least I think that seeing `sel` as a list in the prompt is weird for non-plugin-dev users.
> 
> 
> 
> @TerminalFi What do you think?

Actually, I think that we are missing part of the payload. I need to test it. 

But I actually think it expects the code to be part of the references already. And then after the /review or prompt it expects either blank, or additional message from user. I'll post example here shortly

**Comment by TerminalFi** [comment]
@jfcherng 

I haven't tested, but I think it might expect us to pass in a single reference

```
                    # References
                    # {
                    #     "type": "file", # Literal["file"]
                    #     "status": "" # Literal["included", "blocked", "notfound", "empty"]
                    #     "range": {} # CopilotPayloadCompletionRange,
                    #     "uri": ni.Type.String(),
                    #     "position": ni.Type.Optional(
                    #     "    ni".Type.Object({
                    #     "        line": ni.Type.Number({ minimum: 0 }),
                    #     "        character": ni.Type.Number({ minimum: 0 }),
                    #     "    "}),
                    #     ")",
                    #     "visibleRange": ni.Type.Optional(C2), # CopilotPayloadCompletionRange
                    #     "selection": ni.Type.Optional(C2), # CopilotPayloadCompletionRange
                    #     "openedAt": ni.Type.Optional(ni.Type.String()),
                    #     "activeAt": ni.Type.Optional(ni.Type.String()),
                    # }
```

**Comment by jfcherng** [comment]
you mean the server will auto append codes if the template is built-in?

**Comment by TerminalFi** [comment]
> you mean the server will auto append codes if the template is built-in so appending it again would cause duplication?

Yes, I think that when we send a command like 

`/review` If it contains no `code` snippet the service will then look at the first `ref` which contains the above details, such as visibleRange, Selection, and attempt to use those. And then if the user send `/review {{ code[0] }}` it will use the provided code instead of the ref.

Untested

**Comment by jfcherng** [comment]
Ah yes, if there is no selected code for `/fix`, Copilot just say there is no code.

**Comment by TerminalFi** [comment]
> Ah yes, if there is no selected code for `/fix`, Copilot just say there is no code. We can decide what is expected to happen however.
> 
> some cases to be dealt with:
> 
> * No caret
 
 Assume full file + visiblerange

> * 1 caret, no selected code

 Assume full file + visiblerange

> * 3 carets, only the 2nd has selected code (the 1st and 3rd has no selected code)

1st 3rd are dropped

**Comment by jfcherng** [comment]
> > * 3 carets, only the 2nd has selected code (the 1st and 3rd has no selected code)
> 
> 1st 3rd are dropped

what if 2 carets, both have codes?

**Comment by TerminalFi** [comment]
Also, I have a feeling, if we implement this, we can likely hack together `inline` which I believe Copilot then response with WorkspaceEdits.

But the above statement has a lot of assumptions

**Comment by TerminalFi** [comment]
> Also, I have a feeling, if we implement this, we can likely hack together `inline` which I believe Copilot then response with WorkspaceEdits.
> 
> 
> 
> But the above statement has a lot of assumptions
> 
> 

Since they support multiple refs, let's test sending each non empty carets code snippet as refs

**Comment by TerminalFi** [comment]
FWIW, JetBrains might show more details. It's also using this same node binary

**Comment by TerminalFi** [comment]
> FWIW, JetBrains might show more details. It's also using this same node binary 

@jfcherng sre you working this? Have any local changes ?

**Comment by jfcherng** [comment]
> > FWIW, JetBrains might show more details. It's also using this same node binary 
> 
> @jfcherng sre you working this? Have any local changes ?

no

File: plugin/helpers.py [hunk 1]
```diff
@@ -214,12 +214,13 @@ def prepare_conversation_turn_request(
...
         if selection.empty() or view.substr(selection).strip() == "":
             continue
+        file_path = view.file_name() or f"buffer:{view.buffer().id()}"
         selection_start = view.rowcol(selection.begin())
         selection_end = view.rowcol(selection.end())
         turn["references"].append({
             "type": "file",
             "status": "included",
-            "uri": filename_to_uri(view.file_name()),
+            "uri": file_path if file_path.startswith("buffer:") else file_path and filename_to_uri(file_path),
```
Comment (jfcherng) [inline plugin/helpers.py:223]: is that `file_path and` required?

File: plugin/helpers.py [hunk 2]
```diff
@@ -210,18 +257,33 @@ def preprocess_message_for_html(message: str) -> str:
...
-    if is_template:
-        message += " {{ sel[0] }}"
+    templates = templates or []
+    user_template = first_true(templates, pred=lambda t: f"/{t['id']}" == message)
+    is_template = False
+
+    if user_template:
+        is_template = True
+        message += "\n\n{{ user_prompt }}\n\n{{ code }}"
+        message += "\n\n{{ user_prompt }}\n\n"
```
Comment (jfcherng) [inline plugin/helpers.py:274]: weird duplication?

File: plugin/types.py [hunk 3]
```diff
@@ -180,6 +198,25 @@ class CopilotPayloadConversationTemplate(TypedDict, total=True):
     scopes: list[str]
 
 
+class CopilotRequestCoversationTurn(TypedDict, total=True):
```
Comment (TerminalFi) [inline plugin/types.py:201]: Typo in `Coversation` -> `Conversartion`

File: plugin/types.py [hunk 4]
```diff
@@ -180,6 +198,25 @@ class CopilotPayloadConversationTemplate(TypedDict, total=True):
...
+class CopilotRequestCoversationTurnReference(TypedDict, total=True):
+    type: str
+    status: str
+    uri: str
+    range: CopilotPositionType
+    visibleRange: CopilotRangeType
+    selection: CopilotRangeType
+
+
 class CopilotRequestCoversationAgent(TypedDict, total=True):
```
Comment (TerminalFi) [inline plugin/types.py:220]: Guess I made a lot of typos :D Typo in `Coversation` -> `Conversation`

**Review by TerminalFi — COMMENTED** [review]
Few typos

**Comment by TerminalFi** [comment]
Unrelated to this MR. I noticed all the Copilot Chat stuff stays in command palette when signed out

**Comment by TerminalFi** [comment]
LGTM. We can refine in later MR's after feature lives for awhile

**Review by jfcherng — APPROVED** [review]

**Comment by jfcherng** [comment]
supposedly these branches are no longer used?

-   fix/deleteTurn
-   fixup/Templates

**Comment by TerminalFi** [comment]
> supposedly these branches are no longer used?
> 
> * fix/deleteTurn
> * fixup/Templates

deleted

**Pull request merged**